File: imu_calib_matlab/src/tasks/run_allan_analysis.m
function task = run_allan_analysis(inputData, varargin)
%RUN_ALLAN_ANALYSIS Run Allan deviation analysis using only the available static signals.
% Minimum input:
%   static.t and at least one of static.gyro / static.acc

opts = parse_options(varargin{:});
staticData = normalize_static_input(inputData);
missing = require_inputs(staticData, {'t'});
if ~isempty(missing)
    task = make_task_result(false, 'static.t is required for Allan analysis.', [], ...
        'missing_inputs', {'static.t'}, ...
        'meta', struct('task_name', 'run_allan_analysis'));
    return;
end

hasGyro = isfield(staticData, 'gyro') && ~isempty(staticData.gyro);
hasAcc = isfield(staticData, 'acc') && ~isempty(staticData.acc);
if ~hasGyro && ~hasAcc
    task = make_task_result(false, 'At least one of static.gyro or static.acc is required for Allan analysis.', [], ...
        'missing_inputs', {'static.gyro or static.acc'}, ...
        'meta', struct('task_name', 'run_allan_analysis'));
    return;
end

result = struct();
warnings = {};
if hasGyro
    result.gyro_allan = analyze_gyro_allan(staticData.t, staticData.gyro, varargin{:});
else
    result.gyro_allan = [];
    warnings{end + 1} = 'static.gyro not provided; gyro Allan analysis skipped.'; %#ok<AGROW>
end
if hasAcc
    result.acc_allan = analyze_acc_allan(staticData.t, staticData.acc, varargin{:});
else
    result.acc_allan = [];
    warnings{end + 1} = 'static.acc not provided; accelerometer Allan analysis skipped.'; %#ok<AGROW>
end

task = make_task_result(true, 'Allan analysis completed.', result, ...
    'warnings', warnings, ...
    'meta', struct('task_name', 'run_allan_analysis'));
end

function opts = parse_options(varargin) %#ok<DEFNU>
opts = varargin; %#ok<NASGU>
end

function staticData = normalize_static_input(inputData)
if isstruct(inputData) && isfield(inputData, 'static')
    staticData = inputData.static;
else
    staticData = inputData;
end
end

File: imu_calib_python/imu_calib/tasks/run_allan_analysis.py
from __future__ import annotations

from dataclasses import is_dataclass

from imu_calib.calib.analyze_acc_allan import analyze_acc_allan
from imu_calib.calib.analyze_gyro_allan import analyze_gyro_allan
from imu_calib.tasks.make_task_result import make_task_result
from imu_calib.tasks.require_inputs import require_inputs


def run_allan_analysis(input_data, **kwargs) -> dict:
    static = _normalize_static_input(input_data)
    missing = require_inputs(static, ["t"])
    if missing:
        return make_task_result(
            False,
            "static.t is required for Allan analysis.",
            None,
            missing_inputs=["static.t"],
            meta={"task_name": "run_allan_analysis"},
        )

    has_gyro = hasattr(static, "gyro") and static.gyro is not None and len(static.gyro) > 0
    has_acc = hasattr(static, "acc") and static.acc is not None and len(static.acc) > 0
    if not has_gyro and not has_acc:
        return make_task_result(
            False,
            "At least one of static.gyro or static.acc is required for Allan analysis.",
            None,
            missing_inputs=["static.gyro or static.acc"],
            meta={"task_name": "run_allan_analysis"},
        )

    result = {"gyro_allan": None, "acc_allan": None}
    warnings: list[str] = []
    if has_gyro:
        result["gyro_allan"] = analyze_gyro_allan(static.t, static.gyro, **kwargs)
    else:
        warnings.append("static.gyro not provided; gyro Allan analysis skipped.")
    if has_acc:
        result["acc_allan"] = analyze_acc_allan(static.t, static.acc, **kwargs)
    else:
        warnings.append("static.acc not provided; accelerometer Allan analysis skipped.")

    return make_task_result(
        True,
        "Allan analysis completed.",
        result,
        warnings=warnings,
        meta={"task_name": "run_allan_analysis"},
    )


def _normalize_static_input(input_data):
    if is_dataclass(input_data) and hasattr(input_data, "static"):
        return input_data.static
    if isinstance(input_data, dict) and "static" in input_data:
        return input_data["static"]
    return input_data
